test/test_helpers.jl

```diff
@@ -1,3 +1,100 @@
+using JLArrays  # provides a fake GPU array
+JLArrays.allowscalar(false)
+
+using Adapt
+jl32(xs) = adapt(JLArray{Float32}, xs)
+Adapt.adapt_storage(::Type{<:JLArray{Float32}}, xs::AbstractArray{<:Complex}) = convert(JLArray{ComplexF32}, xs)
+Adapt.adapt_storage(::Type{<:JLArray{Float32}}, x::Float64) = Float32(x)
+Adapt.adapt_storage(::Type{<:JLArray{Float32}}, x::ComplexF64) = ComplexF32(x)
+
+f32(xs) = adapt(Array{Float32}, xs)  # these adapt methods are piracy, just for testing though!
+Adapt.adapt_storage(::Type{<:Array{Float32}}, xs::AbstractArray{<:Complex}) = convert(Array{ComplexF32}, xs)
+Adapt.adapt_storage(::Type{<:Array{Float32}}, x::Float64) = Float32(x)
+Adapt.adapt_storage(::Type{<:Array{Float32}}, x::ComplexF64) = ComplexF32(x)
+
+Adapt.adapt(T::Type{<:JLArray{Float32}}, x::AbstractThunk) = adapt(T, unthunk(x))
+Adapt.adapt(T::Type{<:Array{Float32}}, x::AbstractThunk) = adapt(T, unthunk(x))
+    
+
+"""
+    @gpu test_rrule(sum, rand(3))
+    @gpu test_frule((0, rand(3)), sum, rand(3))
+
+After running the test shown, this converts all arrays to `GPUArray`s,
+and checks that the rule accepts this, and produces the same result.
+
+  @gpu rrule(sum, rand(3))
+  @gpu frule((0, rand(3)), sum, rand(3))
+  
+Used directly on a rule, this compares a CPU to a GPU run, without
+checking correctness. Both should use `Float32`.
+"""
+macro gpu(ex)
+    _gpu_macro(ex, false, __source__)
+end
+macro gpu_broken(ex)
+    _gpu_macro(ex, true, __source__)
+end 
+function _gpu_macro(ex, broken::Bool, __source__)
+    Meta.isexpr(ex, :call) && ex.args[1] in (:test_rrule, :test_frule, :rrule, :frule) ||
+      error("@gpu doesn't understand this input")
+    ex2 = if Meta.isexpr(ex.args[2], :parameters)
+        :($_gpu_test($(ex.args[2]), $(ex.args[1]), $(ex.args[3:end]...)))
+    else
+        :($_gpu_test($(ex.args[1]), $(ex.args[2:end]...)))
+    end
+    return if broken
+        Expr(:block, ex, Expr(:macrocall, Symbol("@test_broken"), __source__, ex2)) |> esc
+    else
+        Expr(:block, ex, Expr(:macrocall, Symbol("@test"), __source__, ex2)) |> esc
+        # NB this @test should report a line number in your tests, not here.
```

Comment:
we really need to document how to make macros do that.
Here is one of the ways
https://github.com/JuliaLogging/LoggingExtras.jl/blob/be221f3147856a7fb2c2dcbd8ea202610a65725d/src/verbosity.jl#L1-L11